File: download-portal-php/includes/captcha.php
<?php
/**
 * Captcha Functions
 */

require_once __DIR__ . '/../config/database.php';
require_once __DIR__ . '/functions.php';

/**
 * Generate math captcha
 */
function generateCaptcha() {
    cleanExpiredCaptchas();
    
    $db = getDB();
    
    $num1 = rand(1, 20);
    $num2 = rand(1, 20);
    
    $operators = [
        ['+', $num1 + $num2],
        ['-', abs($num1 - $num2)],
        ['×', ($num1 < 10 && $num2 < 10) ? $num1 * $num2 : $num1 + $num2]
    ];
    
    $selected = $operators[array_rand($operators)];
    $operator = $selected[0];
    $answer = $selected[1];
    
    // Make subtraction always positive
    if ($operator === '-' && $num1 < $num2) {
        list($num1, $num2) = [$num2, $num1];
        $answer = $num1 - $num2;
    }
    
    $id = generateUUID();
    $expiresAt = date('Y-m-d H:i:s', strtotime('+5 minutes'));
    
    $stmt = $db->prepare("
        INSERT INTO captchas (id, num1, num2, operator, answer, expires_at)
        VALUES (?, ?, ?, ?, ?, ?)
    ");
    $stmt->execute([$id, $num1, $num2, $operator, $answer, $expiresAt]);
    
    return [
        'id' => $id,
        'challenge' => "$num1 $operator $num2 = ?",
        'expires_at' => $expiresAt
    ];
}

/**
 * Verify math captcha
 */
function verifyCaptcha($captchaId, $answer) {
    if (empty($captchaId) || !is_numeric($answer)) {
        return false;
    }
    
    $db = getDB();
    
    $stmt = $db->prepare("SELECT * FROM captchas WHERE id = ? AND expires_at > NOW()");
    $stmt->execute([$captchaId]);
    $captcha = $stmt->fetch();
    
    if (!$captcha) {
        return false;
    }
    
    // Delete the captcha (one-time use)
    $stmt = $db->prepare("DELETE FROM captchas WHERE id = ?");
    $stmt->execute([$captchaId]);
    
    return (int)$answer === (int)$captcha['answer'];
}

/**
 * Verify Google reCAPTCHA v2
 */
function verifyRecaptcha($token, $secretKey) {
    if (empty($token) || empty($secretKey)) {
        return false;
    }
    
    $url = 'https://www.google.com/recaptcha/api/siteverify';
    $data = [
        'secret' => $secretKey,
        'response' => $token,
        'remoteip' => getClientIP()
    ];
    
    $options = [
        'http' => [
            'header' => "Content-type: application/x-www-form-urlencoded\r\n",
            'method' => 'POST',
            'content' => http_build_query($data)
        ]
    ];
    
    $context = stream_context_create($options);
    $result = file_get_contents($url, false, $context);
    
    if ($result === false) {
        return false;
    }
    
    $response = json_decode($result, true);
    return $response['success'] ?? false;
}

/**
 * Verify captcha based on settings
 */
function verifyCaptchaOrRecaptcha($data, $settings, $type = 'submit') {
    $enableKey = ($type === 'auth') ? 'recaptcha_enable_auth' : 'recaptcha_enable_submit';
    
    if ($settings[$enableKey]) {
        // reCAPTCHA enabled
        if (empty($settings['recaptcha_site_key']) || empty($settings['recaptcha_secret_key'])) {
            return ['valid' => false, 'error' => 'reCAPTCHA is enabled but not configured'];
        }
        
        $token = $data['recaptcha_token'] ?? '';
        if (!verifyRecaptcha($token, $settings['recaptcha_secret_key'])) {
            return ['valid' => false, 'error' => 'Invalid reCAPTCHA. Please try again.'];
        }
    } else {
        // Math captcha
        $captchaId = $data['captcha_id'] ?? '';
        $captchaAnswer = $data['captcha_answer'] ?? '';
        
        if (!verifyCaptcha($captchaId, $captchaAnswer)) {
            return ['valid' => false, 'error' => 'Invalid captcha. Please try again.'];
        }
    }
    
    return ['valid' => true];
}

File: backend/services/captcha.py
"""Captcha service"""
import random
import httpx
import logging
from datetime import datetime, timezone, timedelta
from typing import Optional

from services.database import db
from services.email import fetch_site_settings
from models.schemas import Captcha

logger = logging.getLogger(__name__)


async def generate_captcha_challenge() -> dict:
    """Generate a simple math captcha"""
    num1 = random.randint(1, 20)
    num2 = random.randint(1, 20)
    operators = [('+', num1 + num2), ('-', abs(num1 - num2)), ('×', num1 * num2 if num1 < 10 and num2 < 10 else num1 + num2)]
    operator, answer = random.choice(operators)
    
    # Make subtraction always positive
    if operator == '-':
        if num1 < num2:
            num1, num2 = num2, num1
        answer = num1 - num2
    
    captcha = Captcha(
        num1=num1,
        num2=num2,
        operator=operator,
        answer=answer,
        expires_at=(datetime.now(timezone.utc) + timedelta(minutes=5)).isoformat()
    )
    
    await db.captchas.insert_one(captcha.model_dump())
    
    return {
        "id": captcha.id,
        "challenge": f"{num1} {operator} {num2} = ?",
        "expires_at": captcha.expires_at
    }


async def verify_recaptcha(token: str, remote_ip: Optional[str], secret_key: str) -> bool:
    """Verify Google reCAPTCHA v2 token"""
    if not secret_key:
        return False

    async with httpx.AsyncClient() as client:
        resp = await client.post(
            "https://www.google.com/recaptcha/api/siteverify",
            data={"secret": secret_key, "response": token, "remoteip": remote_ip or ""}
        )
        data = resp.json()
        return data.get("success", False)


async def verify_captcha(captcha_id: str, answer: int) -> bool:
    """Verify math captcha answer"""
    captcha = await db.captchas.find_one({"id": captcha_id})
    if not captcha:
        return False
    
    # Check expiration
    expires_at = datetime.fromisoformat(captcha["expires_at"].replace('Z', '+00:00'))
    if datetime.now(timezone.utc) > expires_at:
        await db.captchas.delete_one({"id": captcha_id})
        return False
    
    # Verify answer
    if captcha["answer"] == answer:
        await db.captchas.delete_one({"id": captcha_id})
        return True
    
    return False


async def get_captcha_method():
    """Get current captcha method based on settings"""
    settings = await fetch_site_settings()
    if settings.get("recaptcha_enable_submit") and settings.get("recaptcha_site_key"):
        return "recaptcha"
    return "math"
